framework/src/utils/MooseEnumBase.C

```diff
@@ -127,3 +105,96 @@ MooseEnumBase::checkDeprecatedBase(const std::string & name_upper) const
+    mooseError("You cannot place whitespace around the '=' character in MooseEnumBase");
+
+  // Split on equals sign
+  std::vector<std::string> name_value;
+  MooseUtils::tokenize(MooseUtils::trim(raw_name), name_value, 1, "=");
+
+  // There should be one or two items in the name_value
+  if (name_value.size() < 1 || name_value.size() > 2)
+    mooseError("Invalid option supplied in MooseEnumBase: ", raw_name);
+
+  // Remove un-wanted space around string
+  name_value[0] = MooseUtils::trim(name_value[0]);
+
+  // See if there is a value supplied for this option
+  // strtol allows for proper conversions of both int and hex strings
+  int value;
+  if (name_value.size() == 2)
+    value = strtol(name_value[1].c_str(), NULL, 0);
+  else if (_ids.empty())
+    value = 0;
+  else
+    value = *_ids.rbegin() + 1;
+
+  addEnumerationName(name_value[0], value);
+}
+
+void
+MooseEnumBase::addEnumerationName(const std::string & name, const int & value)
+{
+  // preserve case for raw options, append to list
+  if (!_raw_names.empty())
+    _raw_names += " ";
+  _raw_names += name;
+
+  // convert name to uppercase
+  std::string upper(name);
+  std::transform(upper.begin(), upper.end(), upper.begin(), ::toupper);
+
+  // check ids and names
+  if (_ids.find(value) != _ids.end())
+    mooseError("The id ", value, " already exists in the enumeration.");
+  if (_name_to_id.find(upper) != _name_to_id.end())
+    mooseError("The name ", upper, " already exists in the enumeration.");
+
+  // populate internal datastructures
+  _names.push_back(upper);
+  _ids.insert(value);
+  _name_to_id[upper] = value;
+  _id_to_name[value] = upper;
+  _name_to_raw_name[upper] = name + " "; // add space to make the name unique for removal
+}
+
+int
+MooseEnumBase::id(std::string name) const
+{
+  std::transform(name.begin(), name.end(), name.begin(), ::toupper);
+  auto iter = _name_to_id.find(name);
+  if (iter == _name_to_id.end())
+    mooseError("The name ", name, " is not a possible enumeration value.");
+  return iter->second;
+}
+
+const std::string &
+MooseEnumBase::name(const int & id) const
+{
+  auto iter = _id_to_name.find(id);
+  if (iter == _id_to_name.end())
+    mooseError("The id ", id, " is not a possible enumeration id.");
+  return iter->second;
+}
+
+void
+MooseEnumBase::removeEnumerationName(std::string name)
+{
+  std::transform(name.begin(), name.end(), name.begin(), ::toupper);
+  std::vector<std::string>::iterator iter = std::find(_names.begin(), _names.end(), name);
+  if (iter == _names.end())
+    mooseError("The name ", name, " is not a possible enumeration value, thus can not be removed.");
+  else
+  {
+    _ids.erase(_ids.find(_name_to_id[name]));
+    _name_to_id.erase(_name_to_id.find(name));
+    _names.erase(iter);
+    _raw_names.erase(_raw_names.find(_name_to_raw_name[name]), name.size() + 1);
+    _name_to_raw_name.erase(_name_to_raw_name.find(name));
```

Comment:
What about `_id_to_name`? All these additions to MooseEnum makes we wonder if this is really the right mechanism for this capability. It's a lot of new code to support one new feature.